http-client/src/main/java/io/airlift/http/client/GatheringByteArrayInputStream.java

```diff
@@ -0,0 +1,75 @@
+package io.airlift.http.client;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Iterator;
+import java.util.List;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.lang.Math.min;
+import static java.lang.System.arraycopy;
+import static java.util.Objects.requireNonNull;
+
+public class GatheringByteArrayInputStream
+        extends InputStream
+{
+    private final Iterator<byte[]> arrays;
+    private int remainingBytes;
+    private byte[] currentArray;
+    private int nextByte;
+
+    public GatheringByteArrayInputStream(List<byte[]> arrays, int totalBytes)
+    {
+        checkArgument(totalBytes >= 0, "totalBytes should equal to or greater than 0");
+        this.arrays = requireNonNull(arrays, "arrays is null").iterator();
+        this.remainingBytes = totalBytes;
+    }
+
+    @Override
+    public synchronized int read(byte[] buffer, int offset, int length)
+    {
+        requireNonNull(buffer, "buffer is null");
+        checkArgument(offset >= 0, "offset should equal to or greater than 0");
+        checkArgument(length >= 0, "length should equal to or greater than 0");
+
+        int totalReadBytes = 0;
+        while (totalReadBytes < length && remainingBytes > 0 && (currentArray != null || arrays.hasNext())) {
+            if (currentArray == null) {
+                currentArray = arrays.next();
+                nextByte = 0;
+            }
+            int readBytes = min(remainingBytes, min(length - totalReadBytes, currentArray.length - nextByte));
+            arraycopy(currentArray, nextByte, buffer, offset + totalReadBytes, readBytes);
+            totalReadBytes += readBytes;
+            nextByte += readBytes;
+            remainingBytes -= readBytes;
+
+            // If current array is finished, set current array to null
+            if (nextByte >= currentArray.length || remainingBytes <= 0) {
+                currentArray = null;
+            }
+        }
+
+        return totalReadBytes > 0 ? totalReadBytes : -1;
```

Comment:
This is invalid.  The user can request a read of zero bytes, and in that case you should return `0`. The `-1` result means end of stream.